File: _includes/sort-benchmark/algorithms/counting.rs
fn counting_sort(a: &mut [usize]) {
    if a.is_empty() {
        return;
    }

    let min = *a.iter().min().unwrap();
    let max = *a.iter().max().unwrap();
    let span = max - min + 1;
    let mut count = vec![0usize; span];

    for &x in a.iter() {
        count[x - min] += 1;
    }

    let mut idx = 0;

    for (offset, &cnt) in count.iter().enumerate() {
        let value = min + offset;
        for _ in 0..cnt {
            a[idx] = value;
            idx += 1;
        }
    }
}

File: _includes/sort-benchmark/algorithms/counting.swift
func counting_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { counting_sort($0) }
}

func counting_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    if a.isEmpty {
        return
    }

    var min = a[0]
    var max = a[0]
    for i in 1..<a.count {
        if a[i] < min { min = a[i] }
        if a[i] > max { max = a[i] }
    }
    let span = max - min + 1
    var count = [Int](repeating: 0, count: span)

    for i in 0..<a.count {
        count[a[i] - min] += 1
    }

    var idx = 0
    for offset in 0..<count.count {
        let value = min + offset
        let cnt = count[offset]
        for _ in 0..<cnt {
            a[idx] = value
            idx += 1
        }
    }
}
